### PR TITLE
Test sshkeygen import

### PR DESCRIPTION
**Pull request by williamcroberts**
@niooss-ledger

### CODE REVIEW

**Comment by williamcroberts** [comment]
I had a feeling it was going to fail with this:
```
AttributeError: module 'cryptography.hazmat.primitives.serialization' has no attribute 'load_ssh_private_key'
```

I'll have to retrigger the containers to get newer versions of cryptgoraphy.

File: test/integration/pkcs11-tool.sh [hunk 1]
```diff
@@ -134,4 +134,27 @@ pkcs11_tool --sign --login --slot=1 --id="6d79727361333037326b6579" --pin="myuse
...
+# pkcs11-tool --label doesn't always work, so use id's which are hex encoded.
+echo "testdata">${tempdir}/data
+pkcs11_tool --sign --login --token-label="import-keys" --id="696d706f727465645f7373685f7273615f6b6579" --pin="anotheruserpin" \
+            --input-file ${tempdir}/data --output-file ${tempdir}/sig \
+            --mechanism SHA256-RSA-PKCS
+
+size="$(stat --printf="%s" ${tempdir}/sig)"
+test "$size" -eq "256"
+
+echo "BILL2"
```
Comment (niooss-ledger) [inline test/integration/pkcs11-tool.sh:151]: This `echo "BILL2"` sounds like something you forgot to remove ;)

File: test/integration/scripts/create_pkcs_store.sh [hunk 2]
```diff
@@ -176,6 +176,12 @@ echo "importing ECCp256 key under token 'import-keys'"
 openssl ecparam -name prime256v1 -genkey -noout -out "$TPM2_PKCS11_STORE/private.ecc.pem"
 tpm2_ptool import --privkey="$TPM2_PKCS11_STORE/private.ecc.pem" --algorithm=ecc --key-label="imported_ecc_key" --label="import-keys" --userpin=anotheruserpin --path=$TPM2_PKCS11_STORE
 
+# import an ECC and RSA key in the ssh-keygen format
+ssh-keygen -t rsa -b 2048 -f "$TPM2_PKCS11_STORE/id_rsa_pass" -N 'secret'
+ssh-keygen -t ecdsa -b 256 -f "$TPM2_PKCS11_STORE/id_ec_nopass" -N ''
+tpm2_ptool import --privkey="$TPM2_PKCS11_STORE/id_rsa_pass" --key-label="imported_ssh_rsa_key" --id='imported_ssh_rsa_key' --label="import-keys" --userpin=anotheruserpin --passin 'pass:secret' --path=$TPM2_PKCS11_STORE
+tpm2_ptool import --privkey="$TPM2_PKCS11_STORE/id_ec_nopass" --key-label="imported_ssh_ecc_key" --id='imported_ssh_ecc_key' --label="import-keys" --userpin=anotheruserpin --path=$TPM2_PKCS11_STORE
```
Comment (niooss-ledger) [inline test/integration/scripts/create_pkcs_store.sh:183]: Maybe `--passin env:...` and `--passin file:...` could also be tested?

**Review by niooss-ledger — COMMENTED** [review]
Hello,
Thanks for the tests! Indeed `load_ssh_private_key` was introduced in cryptography 3.0 (according to its documentation, https://cryptography.io/en/3.0/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.load_ssh_private_key) and this version is only packaged since Ubuntu 20.10,  Debian 11 (current "Debian testing"), Fedora 33... (cf. https://repology.org/project/python:cryptography/versions).

Moreover, to be able to decrypt encrypted RSA keys, Python's bcrypt library also needs to be installed. Otherwise `test/integration/pkcs11-tool.sh` fails with:
```text
  File "/usr/lib/python3.9/site-packages/cryptography/hazmat/primitives/serialization/ssh.py", line 39, in _bcrypt_kdf
    raise UnsupportedAlgorithm("Need bcrypt module")
cryptography.exceptions.UnsupportedAlgorithm: Need bcrypt module
Failed to run the setup script
Execution of ./test/integration/pkcs11-tool.sh failed: 1
```
This library could be documented as an optional dependency of `tpm2_ptool`.

With this, the tests look good! I successfully ran them in an Arch Linux container, though with some difficulties to get my environment right. More precisely, here are the commands I ran:
```sh
# Launch a container
podman run --rm -v "$(pwd):/git" --workdir /git -ti docker.io/library/archlinux bash

# Install dependencies
pacman -Sy base-devel autoconf-archive git libyaml tpm2-abrmd tpm2-tools tpm2-tss-engine cmocka expect nss \
    opensc python-python-pkcs11 swtpm jdk-openjdk junit dbus python-yaml python-cryptography \
    python-pyasn1-modules libp11 openssh python-bcrypt vim
export CLASSPATH=/usr/share/java/junit.jar:/usr/share/java/hamcrest-core.jar
mkdir /etc/pkcs11/modules
echo "module: /git/src/.libs/libtpm2_pkcs11.so" >/etc/pkcs11/modules/tpm2_pkcs11.module

# Launch DBus
mkdir -p /run/dbus && dbus-daemon --system --fork

# Build and test tpm2-pkcs11
./bootstrap
./configure --enable-esapi-session-manage-flags --enable-fapi --enable-unit --enable-integration
make
make check
```

Maybe these commands could be added to the building documentation (https://github.com/tpm2-software/tpm2-pkcs11/blob/1.6.0/docs/BUILDING.md) in order to help contributors figuring out how to run tests.

File: test/integration/scripts/create_pkcs_store.sh (hunk 2 above)
Comment (williamcroberts) [reply test/integration/scripts/create_pkcs_store.sh:183]: yeah, right now I am trying to get our containers right for building just to get this to pass, then we can add.

**Comment by williamcroberts** [comment]
> Hello,
> Thanks for the tests! Indeed `load_ssh_private_key` was introduced in cryptography 3.0 (according to its documentation, https://cryptography.io/en/3.0/hazmat/primitives/asymmetric/serialization/#cryptography.hazmat.primitives.serialization.load_ssh_private_key) and this version is only packaged since Ubuntu 20.10, Debian 11 (current "Debian testing"), Fedora 33... (cf. https://repology.org/project/python:cryptography/versions).
> 
> Moreover, to be able to decrypt encrypted RSA keys, Python's bcrypt library also needs to be installed. Otherwise `test/integration/pkcs11-tool.sh` fails with:
> 
> ```
>   File "/usr/lib/python3.9/site-packages/cryptography/hazmat/primitives/serialization/ssh.py", line 39, in _bcrypt_kdf
>     raise UnsupportedAlgorithm("Need bcrypt module")
> cryptography.exceptions.UnsupportedAlgorithm: Need bcrypt module
> Failed to run the setup script
> Execution of ./test/integration/pkcs11-tool.sh failed: 1
> ```
> 
> This library could be documented as an optional dependency of `tpm2_ptool`.

Lets just make it mandatory, cryptography >= 3.0 is pretty old. Considering pip and you're not dependent on distro packagers, it's easy enough to fulfill IMO. But I guess for packager it might be harder. We also need to add bcrypt to setup.py, and probably port that to a setup.cfg, and probably also test that it works.

> 
> With this, the tests look good! I successfully ran them in an Arch Linux container, though with some difficulties to get my environment right. More precisely, here are the commands I ran:
> 
> ```shell
> # Launch a container
> podman run --rm -v "$(pwd):/git" --workdir /git -ti docker.io/library/archlinux bash
> 
> # Install dependencies
> pacman -Sy base-devel autoconf-archive git libyaml tpm2-abrmd tpm2-tools tpm2-tss-engine cmocka expect nss \
>     opensc python-python-pkcs11 swtpm jdk-openjdk junit dbus python-yaml python-cryptography \
>     python-pyasn1-modules libp11 openssh python-bcrypt vim
> export CLASSPATH=/usr/share/java/junit.jar:/usr/share/java/hamcrest-core.jar
> mkdir /etc/pkcs11/modules
> echo "module: /git/src/.libs/libtpm2_pkcs11.so" >/etc/pkcs11/modules/tpm2_pkcs11.module
> 
> # Launch DBus
> mkdir -p /run/dbus && dbus-daemon --system --fork
> 
> # Build and test tpm2-pkcs11
> ./bootstrap
> ./configure --enable-esapi-session-manage-flags --enable-fapi --enable-unit --enable-integration
> make
> make check
> ```
> 
> Maybe these commands could be added to the building documentation (https://github.com/tpm2-software/tpm2-pkcs11/blob/1.6.0/docs/BUILDING.md) in order to help contributors figuring out how to run tests.

Patches are always welcome. I am not familiar with those toolsets. But we have a docker container images athttps://github.com/orgs/tpm2-software/packages which hail from the tpm2-software-container dockerfiles.

**Comment by williamcroberts** [comment]
ughh that moment you git foo incorrectly and lose your branch. Thankfully git reflog had them.

**Pull request merged**